File: Python/Data Structures & Algorithms Module - Python/Week2/LinkedLists/LinkedList.py
"""
LinkedList.py
Billy Cussen
10/11/2020
"""
from Node import Node

class LinkedList:

    def __init__(self):
        self.first = None

    def add(self, element):
        if self.isEmpty():
            self.first = Node(element,None)
        else:
            current = self.first
            while current.next != None:
                current = current.next
            current.next = Node(element,None)

    def addByIndex(self, element, index):
        if self.isEmpty():
            self.add(element)
        elif index < 0 or index > self.size():
            raise Exception("ADD BY INDEX: Null Pointer")
        elif index == 0:
            self.first = Node(element,self.first)
        else:
            current = self.first
            for i in range(index-1):
                current = current.next
            current.next = Node(element, current.next)

    def addBefore(self,element, target):
        index = self.getIndex(target)
        self.addByIndex(element,index)

    def addAfter(self,element, target):
        index = self.getIndex(target)
        self.addByIndex(element,index+1)

    def getIndex(self, target):
        current = self.first
        index = 0
        while current != None:
            if(current.element == target):
                return index
            index+=1
            current = current.next
        return -1

    def removeByElement(self, target):
        index = self.getIndex(target)
        return self.removeByIndex(index)

    def removeByIndex(self, index):
        if self.isEmpty() or index < 0 or index > self.size():
            raise Exception("REMOVE BY INDEX: Null Pointer")
        elif index == 0:
            elem = self.first.element
            self.first = self.first.next
            return elem
        else:
            current = self.first
            for i in range(index-1):
                current = current.next
            elem = current.next.element
            current.next = current.next.next
            return elem


    def isEmpty(self):
        return self.first == None

    def size(self):
        return self.sizeHelper(self.first)

    def sizeHelper(self,current):
        if(current == None):
            return 0
        return 1 + self.sizeHelper(current.next)

    def toString(self):
        if self.isEmpty():
            return "The List is Empty!"
        position = 0
        result = "***LINKEDLIST TRAVERSAL***\n"
        return self.toStringHelper(self.first,result,position)

    def toStringHelper(self, current, result, position):
        if current == None:
            result += "***END***"
            return result
        result += "Position "+str(position)+", Element: "+str(current.element)+"\n"
        position+=1
        return self.toStringHelper(current.next,result,position)

File: Java/Data Structures & Algorithms Module - Java/Revision/LinkedList/LinkedList.java
/*
LinkedList.java
Billy Cussen
05/01/2020
*/
public class LinkedList<T> {
    
    Node<T> first;

    public LinkedList(){
        this.first = null;
    }

    public void add(T element){
        if(isEmpty()){
            this.first = new Node<T>(element, null);
        } else{
            Node<T> current = this.first;
            while(current.next!=null){
                current = current.next;
            }
            current.next = new Node<T>(element, null);
        }
    }

    public void addByIndex(T element, int index){
        if(isEmpty() || index < 0 || index > size()){
            throw new NullPointerException("List is Empty!");
        } else if(index == 0){
            this.first = new Node<T>(element, this.first);
        } else{
            Node<T> current = first;
            for(int i = 0; i < index-1;i++){
                current = current.next;
            }
            current.next = new Node<T>(element, current.next);
        }
    }

    public T removeByIndex(int index){
        if(isEmpty() || index < 0 || index > size()){
            throw new NullPointerException("LinkedList is Empty!");
        } else if(index == 0){
            T element = first.element;
            first = first.next;
            return element;
        } else{
            Node<T> current = first;
            for(int i = 0; i < index - 1; i++){
                current = current.next;
            }
            T element = current.next.element;
            current.next = current.next.next;
            return element;
        }
    }

    public boolean isEmpty(){
        return this.first == null;
    }

    public int size(){
        if(isEmpty()){
            throw new NullPointerException("LinkedList is Empty!");
        }
        return sizeHelper(this.first);
    }

    private int sizeHelper(Node<T> current){
        if(current == null){
            return 0;
        }
        return 1 + sizeHelper(current.next);
    }

    @Override
    public String toString(){
        if(isEmpty()){
            throw new NullPointerException("List is Empty!");
        }
        String result = "***LINKEDLIST TRAVERSAL***\n";
        int position = 1;
        return toStringHelper(this.first, result, position);
    }

    private String toStringHelper(Node<T> current, String result, int position){
        if(current == null){
            result+="***END***\n";
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringHelper(current.next, result, ++position);
    }

}
